Handle `--bare` clone targets

hyperrealist opened the pull request:
Fixes #9270

- [x] Append `.git` to target when cloned with `--bare` option
- [x] Write test case

Commit:
append  to bare clone target

andyfeller commented on git/client.go:
```diff
@@ -595,6 +596,9 @@ func (c *Client) Clone(ctx context.Context, cloneURL string, args []string, mods
 	} else {
 		target = path.Base(strings.TrimSuffix(cloneURL, ".git"))
 	}
+	if slices.Contains(cloneArgs, "--bare") {
+		target += ".git"
+	}
```
Seems straightforward 👍

**Do you see any reason a new test case couldn't be added to `TestClientClone`?**  I think this is the only thing needed to get this shipped and merged.

https://github.com/cli/cli/blob/9bd8d44209ebb0a5cfa18f2b796d5feb6f002d65/git/client_test.go#L1267-L1314

andyfeller replied:
@hyperrealist : just letting you know I added this test and pushed it to your branch 👍

andyfeller reviewed (dismissed):
Thanks for reporting this issue and opening the PR to fix it, @hyperrealist!

Changes look good, just want to ensure we capture this in our testing before shipping it.

hyperrealist commented:
Hi @andyfeller, I'll admit that I am a very inexperienced go developer, in fact this is my first time developing in go. I have been scratching my head to understand how `TestClientClone` is structured. If it is a trivial step for you to add a test case, please go ahead and do so.

williammartin commented:
@hyperrealist can you tell me a bit more about your use case for bare repos?

My understanding of bare repos comes from:

> A bare repository might seem to be of little use, but its role is crucial: to serve as an authoritative focal point for collaborative development. Other developers clone and fetch from the bare repository and push updates to it... if you set up a repository into which developers push changes, it should be bare. In effect, this is a special case of the more general best practice that a published repository should be bare.

I'm just wondering whether the set up we do for [bare repos even makes sense](https://github.com/cli/cli/blob/ca1f2a4c48fb84c22ef91994ca6ded011b12d874/pkg/cmd/repo/clone/clone.go#L189-L222) at all?

williammartin commented:
Just FYI I asked a question over on #9270 to double check this is the right direction.

andyfeller commented:
> Just FYI I asked a question over on #9270 to double check this is the right direction.

Good catch on the post clone handling. 🤔 Most of the time I've worked with bare repositories were when I didn't expect to make any changes.

https://github.com/cli/cli/blob/ca1f2a4c48fb84c22ef91994ca6ded011b12d874/pkg/cmd/repo/clone/clone.go#L181-L223

andyfeller committed:
Add bare clone test

williammartin commented on git/client.go:
```diff
@@ -595,6 +596,9 @@ func (c *Client) Clone(ctx context.Context, cloneURL string, args []string, mods
 		target = path.Base(strings.TrimSuffix(cloneURL, ".git"))
 	}
+	if slices.Contains(cloneArgs, "--bare") {
+		target += ".git"
```
I don't know if this is quite right. Just writing a comment here to ensure this isn't merged before I have a moment to investigate.

williammartin replied:
Yes, as I suspected this is broken in the case where you provide both `--bare` **and** a target directory i.e.:

```
➜  cli git:(support-bear) ./bin/gh repo clone hyperrealist/tiled tiled -- --bare
Cloning into bare repository 'tiled'...
remote: Enumerating objects: 16472, done.
remote: Counting objects: 100% (376/376), done.
remote: Compressing objects: 100% (113/113), done.
remote: Total 16472 (delta 283), reused 336 (delta 263), pack-reused 16096
Receiving objects: 100% (16472/16472), 4.30 MiB | 11.09 MiB/s, done.
Resolving deltas: 100% (12482/12482), done.
failed to run git: fatal: cannot change to 'tiled.git': No such file or directory
```

We should only be appending `.git` to the target directory if the user didn't specify one.

andyfeller committed:
Only override implicit bare repo clone target

Follow up to PR feedback from @williammartin, `git clone --bare` will only override the local repository target if unspecified.

This commit enhances the cloning logic and associated test to ensure explicit targets are preserved when cloning.

williammartin commented:
### Acceptance Criteria

**Given** I have a forked repo <repo>
**When** I run `gh repo clone <repo> -- --bare`
**Then** it is cloned and set up correctly with upstream/origin

```
➜  cli git:(support-bear) ./bin/gh repo clone hyperrealist/tiled -- --bare
Cloning into bare repository 'tiled.git'...
remote: Enumerating objects: 16472, done.
remote: Counting objects: 100% (376/376), done.
remote: Compressing objects: 100% (109/109), done.
remote: Total 16472 (delta 283), reused 340 (delta 267), pack-reused 16096
Receiving objects: 100% (16472/16472), 4.30 MiB | 9.59 MiB/s, done.
Resolving deltas: 100% (12478/12478), done.
remote: Enumerating objects: 67, done.
remote: Counting objects: 100% (51/51), done.
remote: Compressing objects: 100% (31/31), done.
remote: Total 67 (delta 25), reused 41 (delta 20), pack-reused 16
Unpacking objects: 100% (67/67), 41.29 KiB | 254.00 KiB/s, done.
From https://github.com/bluesky/tiled
 * [new branch]        main       -> upstream/main
 * [new tag]           v0.1.0a104 -> v0.1.0a104
 * [new tag]           v0.1.0a105 -> v0.1.0a105
 * [new tag]           v0.1.0a106 -> v0.1.0a106
 * [new tag]           v0.1.0a107 -> v0.1.0a107
 * [new tag]           v0.1.0a108 -> v0.1.0a108
 * [new tag]           v0.1.0a109 -> v0.1.0a109
 * [new tag]           v0.1.0a110 -> v0.1.0a110
 * [new tag]           v0.1.0a111 -> v0.1.0a111
 * [new tag]           v0.1.0a113 -> v0.1.0a113
 * [new tag]           v0.1.0a114 -> v0.1.0a114
 * [new tag]           v0.1.0a115 -> v0.1.0a115
 * [new tag]           v0.1.0a116 -> v0.1.0a116
 * [new tag]           v0.1.0a117 -> v0.1.0a117
 * [new tag]           v0.1.0a118 -> v0.1.0a118
 * [new tag]           v0.1.0a119 -> v0.1.0a119
 * [new tag]           v0.1.0a120 -> v0.1.0a120
 * [new tag]           v0.1.0a121 -> v0.1.0a121
 * [new tag]           v0.1.0a122 -> v0.1.0a122
 * [new tag]           v0.1.0b1   -> v0.1.0b1
 * [new tag]           v0.1.0b2   -> v0.1.0b2
 * [new tag]           v0.1.0b3   -> v0.1.0b3
 * [new tag]           v0.1.0b4   -> v0.1.0b4
 * [new tag]           v0.1.0b6   -> v0.1.0b6
 * [new tag]           v0.1.0b5   -> v0.1.0b5
! Repository bluesky/tiled set as the default repository. To learn more about the default repository, run: gh repo set-default --help
➜  cli git:(support-bear) ✗ cd tiled.git && git remote -v
origin     https://github.com/hyperrealist/tiled.git (fetch)
origin  https://github.com/hyperrealist/tiled.git (push)
upstream        https://github.com/bluesky/tiled.git (fetch)
upstream        https://github.com/bluesky/tiled.git (push)
```

**Given** I have a forked repo <repo>
**When** I run `gh repo clone <repo> myrepo -- --bare`
**Then** it is cloned into dir `myrepo` and set up correctly with upstream/origin

```
➜  cli git:(support-bear) ./bin/gh repo clone hyperrealist/tiled myrepo -- --bare
Cloning into bare repository 'myrepo'...
remote: Enumerating objects: 16472, done.
remote: Counting objects: 100% (376/376), done.
remote: Compressing objects: 100% (113/113), done.
remote: Total 16472 (delta 283), reused 336 (delta 263), pack-reused 16096
Receiving objects: 100% (16472/16472), 4.30 MiB | 8.58 MiB/s, done.
Resolving deltas: 100% (12482/12482), done.
remote: Enumerating objects: 67, done.
remote: Counting objects: 100% (51/51), done.
remote: Compressing objects: 100% (31/31), done.
remote: Total 67 (delta 25), reused 41 (delta 20), pack-reused 16
Unpacking objects: 100% (67/67), 41.29 KiB | 291.00 KiB/s, done.
From https://github.com/bluesky/tiled
 * [new branch]        main       -> upstream/main
 * [new tag]           v0.1.0a104 -> v0.1.0a104
 * [new tag]           v0.1.0a105 -> v0.1.0a105
 * [new tag]           v0.1.0a106 -> v0.1.0a106
 * [new tag]           v0.1.0a107 -> v0.1.0a107
 * [new tag]           v0.1.0a108 -> v0.1.0a108
 * [new tag]           v0.1.0a109 -> v0.1.0a109
 * [new tag]           v0.1.0a110 -> v0.1.0a110
 * [new tag]           v0.1.0a111 -> v0.1.0a111
 * [new tag]           v0.1.0a113 -> v0.1.0a113
 * [new tag]           v0.1.0a114 -> v0.1.0a114
 * [new tag]           v0.1.0a115 -> v0.1.0a115
 * [new tag]           v0.1.0a116 -> v0.1.0a116
 * [new tag]           v0.1.0a117 -> v0.1.0a117
 * [new tag]           v0.1.0a118 -> v0.1.0a118
 * [new tag]           v0.1.0a119 -> v0.1.0a119
 * [new tag]           v0.1.0a120 -> v0.1.0a120
 * [new tag]           v0.1.0a121 -> v0.1.0a121
 * [new tag]           v0.1.0a122 -> v0.1.0a122
 * [new tag]           v0.1.0b1   -> v0.1.0b1
 * [new tag]           v0.1.0b2   -> v0.1.0b2
 * [new tag]           v0.1.0b3   -> v0.1.0b3
 * [new tag]           v0.1.0b4   -> v0.1.0b4
 * [new tag]           v0.1.0b6   -> v0.1.0b6
 * [new tag]           v0.1.0b5   -> v0.1.0b5
! Repository bluesky/tiled set as the default repository. To learn more about the default repository, run: gh repo set-default --help
➜  cli git:(support-bear) ✗ cd myrepo && git remote -v
origin      https://github.com/hyperrealist/tiled.git (fetch)
origin  https://github.com/hyperrealist/tiled.git (push)
upstream        https://github.com/bluesky/tiled.git (fetch)
upstream        https://github.com/bluesky/tiled.git (push)
```

LGTM
